internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,263 @@
+		pollerPermitLastUpdate: input.Clock.Now(),
+	}
+}
+
+func (c *ConcurrencyAutoScaler) Start() {
+	c.logEvent(autoScalerEventStart)
+
+	c.wg.Add(1)
+
+	go func() {
+		defer c.wg.Done()
+		ticker := c.clock.NewTicker(c.updateTick)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-c.shutdownChan:
+				return
+			case <-ticker.Chan():
+				c.logEvent(autoScalerEventMetrics)
+				c.updatePollerPermit()
+			}
+		}
+	}()
+}
+
+func (c *ConcurrencyAutoScaler) Stop() {
+	close(c.shutdownChan)
+	c.wg.Wait()
+	c.logEvent(autoScalerEventStop)
+}
+
+// ProcessPollerHint reads the poller response hint and take actions
+// 1. update poller wait time
+// 2. enable/disable auto scaler
+func (c *ConcurrencyAutoScaler) ProcessPollerHint(hint *shared.AutoConfigHint) {
+	if hint == nil {
+		c.log.Warn("auto config hint is nil, this results in no action")
+		return
+	}
+	if hint.PollerWaitTimeInMs != nil {
+		waitTimeInMs := *hint.PollerWaitTimeInMs
+		c.pollerWaitTimeInMsLog2.Add(math.Log2(float64(waitTimeInMs + 1)))
+	}
+
+	/*
+		Atomically compare and switch the auto scaler enable flag. If auto scaler is turned off, IMMEDIATELY reset the concurrency limits.
+	*/
+	var shouldEnable bool
+	if hint.EnableAutoConfig != nil && *hint.EnableAutoConfig {
+		shouldEnable = true
+	}
+	if switched := c.enable.CompareAndSwap(!shouldEnable, shouldEnable); switched {
+		if shouldEnable {
+			c.logEvent(autoScalerEventEnable)
+		} else {
+			c.resetConcurrency()
+			c.logEvent(autoScalerEventDisable)
+		}
+	}
+}
+
+// resetConcurrency reset poller quota to the max value. This will be used for gracefully switching the auto scaler off to avoid workers stuck in the wrong state
+func (c *ConcurrencyAutoScaler) resetConcurrency() {
+	c.concurrency.PollerPermit.SetQuota(c.pollerInitCount)
+}
+
+func (c *ConcurrencyAutoScaler) logEvent(event autoScalerEvent) {
+	if c.enable.Load() {
+		c.scope.Counter("concurrency_auto_scaler.enabled").Inc(1)
+	} else {
+		c.scope.Counter("concurrency_auto_scaler.disabled").Inc(1)
+	}
+	c.scope.Gauge("poller_in_action").Update(float64(c.concurrency.PollerPermit.Count()))
+	c.scope.Gauge("poller_quota").Update(float64(c.concurrency.PollerPermit.Quota()))
+	c.scope.Gauge("poller_wait_time").Update(math.Exp2(c.pollerWaitTimeInMsLog2.Average()))
+	c.log.Debug(autoScalerEventLogMsg,
+		zap.Time("time", c.clock.Now()),
+		zap.String("event", string(event)),
+		zap.Bool("enabled", c.enable.Load()),
+		zap.Int("poller_quota", c.concurrency.PollerPermit.Quota()),
+		zap.Int("poller_in_action", c.concurrency.PollerPermit.Count()),
+	)
+}
+
+func (c *ConcurrencyAutoScaler) updatePollerPermit() {
+	if !c.enable.Load() { // skip update if auto scaler is disabled
+		c.logEvent(autoScalerEventPollerSkipUpdateNotEnabled)
+		return
+	}
+	updateTime := c.clock.Now()
+	if updateTime.Before(c.pollerPermitLastUpdate.Add(c.cooldown)) { // before cooldown
+		c.logEvent(autoScalerEventPollerSkipUpdateCooldown)
+		return
+	}
```

Comment:
since this func (and therefore `updateTime`'s last value) is only ever called in a loop driven by a timer, it seems like this only has two modes of operation:
- cooldown < tick: every tick causes an update
- cooldown > tick: every N ticks cause an update

neither of which seems particularly useful.